### ISSUE TITLE
Do you have any plan to support custom layer?

### ISSUE THREAD

**Issue by chagyun0213** [nvdla/sw#41]
Hi
I believe DLA hw could not cover all operations, for example, like  softmax layer. To resolve this limitation, you had provided Emulation layers. My fundamental question is how to add my custom layer as one of Emulation layer and how to let your nvdla_compiler know this availability to compile my network correctly?
Regards,
CJ

**Comment by prasshantg** [comment]
That requires releasing either full compiler source code or customized compiler. Another way is, you can push change for implementation in runtime and we will add this layer to compiler. What layer do you want to add?

**Comment by chagyun0213** [comment]
Thanks for your reply. I wan to run SSD network - https://github.com/weiliu89/caffe/tree/ssd

I saw the SSD deploy.prototxt file and found there are some custom or not-supported-yet layers. 
Some of them would be supported by nvdla_compiler soon, but some of them not defined by BVLC Caffe might not be. So I am curious to know what's the flow to workaround this kind of custom layers.

I think if your nvdla_compiler do continue compiling deploy.prototxt file not stopping during compilation when it sees unsupported layers, then finally generating C-stubs codes for S/W emulation
would be really  good. In that case, I could code my logic into this stub to complete. I don't know it works nor not. Just idea. 


layer {
  name: "conv4_3_norm"
  type: "Normalize"
  bottom: "conv4_3"
  top: "conv4_3_norm"
  norm_param {
    across_spatial: false
    scale_filler {
      type: "constant"
      value: 20.0
    }
    channel_shared: false
  }
}

layer {
  name: "conv4_3_norm_mbox_loc_perm"
  type: "Permute"
  bottom: "conv4_3_norm_mbox_loc"
  top: "conv4_3_norm_mbox_loc_perm"
  permute_param {
    order: 0
    order: 2
    order: 3
    order: 1
  }
}

layer {
  name: "conv4_3_norm_mbox_conf_flat"
  type: "Flatten"
  bottom: "conv4_3_norm_mbox_conf_perm"
  top: "conv4_3_norm_mbox_conf_flat"
  flatten_param {
    axis: 1
  }
}

layer {
  name: "conv4_3_norm_mbox_priorbox"
  type: "PriorBox"
  bottom: "conv4_3_norm"
  bottom: "data"
  top: "conv4_3_norm_mbox_priorbox"
  prior_box_param {
    min_size: 21.0
    max_size: 45.0
    aspect_ratio: 2.0
    flip: true
    clip: false
    variance: 0.10000000149
    variance: 0.10000000149
    variance: 0.20000000298
    variance: 0.20000000298
    step: 8.0
    offset: 0.5
  }
}

layer {
  name: "mbox_loc"
  type: "Concat"
  bottom: "conv4_3_norm_mbox_loc_flat"
  bottom: "fc7_mbox_loc_flat"
  bottom: "conv6_2_mbox_loc_flat"
  bottom: "conv7_2_mbox_loc_flat"
  bottom: "conv8_2_mbox_loc_flat"
  bottom: "conv9_2_mbox_loc_flat"
  top: "mbox_loc"
  concat_param {
    axis: 1
  }
}

**Comment by prasshantg** [comment]
Yes, it is possible. Please refer to https://github.com/nvdla/sw/blob/master/umd/core/runtime/Emulator.cpp as an example, it implements power and softmax layers. You can implement your own layers in it.

**Comment by aaronduino** [comment]
> it implements power and softmax layers

@prasshantg as a side note, does that mean that power layers are now supported (relevant to #67)?

Either way, I'm glad to hear that it is in fact possible. How should one go about getting the compiler to let the runtime take care of a layer rather than erroring out?

**Comment by honorpeter** [comment]
@prasshantg 
when I do nvdla_compiler for VGG_SSD300 on VP.  like this

sudo ./nvdla_compiler --prototxt deploy_tod5.prototxt --caffemodel VGG_TOD_SSD_300x300_iter_120000.caffemodel
creating new wisdom context...
opening wisdom context...
parsing caffe network...
[libprotobuf ERROR google/protobuf/text_format.cc:274] Error parsing text-format ditcaffe.NetParameter: 891:17: Message type "ditcaffe.LayerParameter" has no field named "permute_param".
Caffe Parser: could not parse text file
libnvdla<1> Could not parse deploy file
(DLA) Error 0x00000004: Unable to parse caffemodel: "deploy_tod5.prototxt" (in ParseTest.cpp, function parseCaffeNetwork(), line 91)
(DLA) Error 0x00000004: (propagating from ParseTest.cpp, function parseAndCompile(), line 164)
(DLA) Error 0x00000004: (propagating from main.cpp, function launchTest(), line 87)

And in my deploy_tod5.prototxt at 891:

layer {
  name: "conv4_3_norm_mbox_loc_perm"
  type: "Permute"
  bottom: "conv4_3_norm_mbox_loc"
  top: "conv4_3_norm_mbox_loc_perm"
  permute_param {
    order: 0
    order: 2
    order: 3
    order: 1
  }
}

Is this layer not-supported-yet layers?
Will you  add this layer to compiler?

**Comment by prasshantg** [comment]
@honorpeter that is right, this layer is not supported currently. Will add to our release planning.
